skunk/rust/skunk/src/main.rs

```diff
@@ -74,24 +76,34 @@ impl MainData {
     }
     Ok(())
   }
-  fn main_module_for_file(&self, location: &str) -> Option<Rc<ast::Module>> {
+  fn get_file_info(&self, location: &str) -> Rc<FileData> {
     let existing_data = {
       let file_info = self.file_info.borrow();
       file_info.get(location).map(|r| (*r).clone())
     };
     match existing_data {
       None => {
-        self.load_file(location).ok()?;
-        self.file_info.borrow().get(location).and_then(|file_info| (file_info.main_module.clone()))    
+        self.load_file(location).ok();
+        let file_info = self.file_info.borrow();
+        file_info.get(location).unwrap().clone()
       }
-      Some(info) => info.main_module.clone()
+      Some(info) => info.clone()
     }
   }
+  fn main_module_for_file(&self, location: &str) -> Option<Rc<ast::Module>> {
+    self.get_file_info(location).main_module.clone()
+  }
+  fn named_module_from_file(&self, location: &str, name: &str) -> Option<Rc<ast::Module>> {
+    self.get_file_info(location).module_map.get(name).map(|rc| rc.clone())
+  }
+  fn named_newtype_from_file(&self, location: &str, name: &str) -> Option<Rc<ast::NewType>> {
+    self.get_file_info(location).newtype_map.get(name).map(|rc| rc.clone())
+  }
 }
 
 impl FileData {
   fn new() -> Self {
-    Self { ast: Vec::new(), main_module: None, buffer: String::new() }
+    Self { ast: Vec::new(), main_module: None, buffer: String::new(), module_map: HashMap::new(), newtype_map: HashMap::new() }
```

Comment:
Nit: This implementation could just call a derived Default:: default(), making it explicit doesn't seem to add much here.

Reply:
I didn't know about Default! This is awesome!